haystack/tools/toolset.py

```diff
@@ -0,0 +1,305 @@
+
+    def __contains__(self, item: Any) -> bool:
+        """
+        Check if a tool is in this Toolset.
+
+        Supports checking by:
+        - Tool instance: tool in toolset
+        - Tool name: "tool_name" in toolset
+
+        :param item: Tool instance or tool name string
+        :returns: True if contained, False otherwise
+        """
+        if isinstance(item, str):
+            return any(tool.name == item for tool in self.tools)
+        if isinstance(item, Tool):
+            return item in self.tools
+        return False
+
+    def add(self, tool: Union[Tool, "Toolset"]) -> None:
+        """
+        Add a new Tool or merge another Toolset.
+
+        :param tool: A Tool instance or another Toolset to add
+        :raises ValueError: If adding the tool would result in duplicate tool names
+        :raises TypeError: If the provided object is not a Tool or Toolset
+        """
+        new_tools = []
+
+        if isinstance(tool, Tool):
+            new_tools = [tool]
+        elif isinstance(tool, Toolset):
+            new_tools = list(tool)
+        else:
+            raise TypeError(f"Expected Tool or Toolset, got {type(tool).__name__}")
+
+        # Check for duplicates before adding
+        combined_tools = self.tools + new_tools
+        _check_duplicate_tool_names(combined_tools)
+
+        self.tools.extend(new_tools)
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize the Toolset to a dictionary.
+
+        :returns: A dictionary representation of the Toolset
+
+        Note for subclass implementers:
+        The default implementation is ideal for scenarios where Tool resolution is static. However, if your subclass
+        of Toolset dynamically resolves Tool instances from external sources—such as an MCP server, OpenAPI URL, or
+        a local OpenAPI specification—you should consider serializing the endpoint descriptor instead of the Tool
+        instances themselves. This strategy preserves the dynamic nature of your Toolset and minimizes the overhead
+        associated with serializing potentially large collections of Tool objects. Moreover, by serializing the
+        descriptor, you ensure that the deserialization process can accurately reconstruct the Tool instances, even
+        if they have been modified or removed since the last serialization. Failing to serialize the descriptor may
+        lead to issues where outdated or incorrect Tool configurations are loaded, potentially causing errors or
+        unexpected behavior.
+        """
+        return {
+            "type": generate_qualified_class_name(type(self)),
+            "data": {"tools": [tool.to_dict() for tool in self.tools]},
+        }
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "Toolset":
+        """
+        Deserialize a Toolset from a dictionary.
+
+        :param data: Dictionary representation of the Toolset
+        :returns: A new Toolset instance
+        """
```

Comment:
maybe I'm wrong but could we perhaps use some of the existing deserialization functions here?